tiledb/sm/query/reader.cc

```diff
@@ -181,15 +176,101 @@ Layout Reader::layout() const {
   return layout_;
 }
 
-void Reader::next_subarray_partition() {
-  if (read_state_.idx_ >= read_state_.subarray_partitions_.size())
-    return;
+Status Reader::next_subarray_partition() {
+  if (read_state_.subarray_partitions_.empty()) {
+    if (read_state_.cur_subarray_ != nullptr) {
+      std::free(read_state_.cur_subarray_);
+      read_state_.cur_subarray_ = nullptr;
+    }
+    return Status::Ok();
+  }
 
-  ++read_state_.idx_;
-  if (read_state_.idx_ >= read_state_.subarray_partitions_.size())
-    return;
+  // Prepare buffer sizes map
+  std::unordered_map<std::string, std::pair<uint64_t, uint64_t>>
+      buffer_sizes_map;
+  for (const auto& it : attr_buffers_) {
+    buffer_sizes_map[it.first] = std::pair<uint64_t, uint64_t>(
+        *(it.second.buffer_size_),
+        (array_schema_->var_size(it.first)) ? *(it.second.buffer_var_size_) :
+                                              0);
+  }
+
+  // Loop until a new partition whose result fit in the buffers is found
+  std::unordered_map<std::string, std::pair<uint64_t, uint64_t>>
+      max_buffer_sizes;
+  bool found = false;
+  void* next_partition = nullptr;
+  auto domain = array_schema_->domain();
+  do {
+    // Pop next partition
+    next_partition = read_state_.subarray_partitions_.front();
```

Comment:
I think this leaks -- you pop this from the list and copy it to `cur_subarray_` but I don't see the popped memory being freed.

Reply:
Good catch!